Lib/test/test_capi/test_codecs.py

```diff
@@ -745,7 +745,69 @@ def test_codec_stream_writer(self):
                 codec_stream_writer(NULL, stream, 'strict')
 
 
+class UnsafeUnicodeEncodeError(UnicodeEncodeError):
+    def __init__(self, encoding, message, start, end, reason):
+        self.may_crash = (end - start) < 0 or (end - start) >= len(message)
+        super().__init__(encoding, message, start, end, reason)
+
+
+class UnsafeUnicodeDecodeError(UnicodeDecodeError):
+    def __init__(self, encoding, message, start, end, reason):
+        # the case end - start >= len(message) does not crash
+        self.may_crash = (end - start) < 0
+        super().__init__(encoding, message, start, end, reason)
+
+
+class UnsafeUnicodeTranslateError(UnicodeTranslateError):
+    def __init__(self, message, start, end, reason):
+        # <= 0 because PyCodec_ReplaceErrors tries to check the Unicode kind
+        # of a 0-length result (which is by convention PyUnicode_1BYTE_KIND
+        # and not PyUnicode_2BYTE_KIND as it currently expects)
+        self.may_crash = (end - start) <= 0 or (end - start) >= len(message)
+        super().__init__(message, start, end, reason)
+
+
 class CAPICodecErrors(unittest.TestCase):
+    @classmethod
+    def _generate_exceptions(cls, atomic_literal, factory, objlens):
+        for objlen in objlens:
+            obj = atomic_literal * objlen
+            m = 2 * max(2, objlen)
+            for start in range(-m, m):
+                for end in range(-m, m):
+                    yield factory(obj, start, end)
+
+    @classmethod
+    def generate_encode_errors(cls, objlen, *objlens):
+        def factory(obj, start, end):
+            return UnsafeUnicodeEncodeError('utf-8', obj, start, end, 'reason')
+        return tuple(cls._generate_exceptions('0', factory, [objlen, *objlens]))
```

Comment:
It might be easier to follow this without passing a `factory` function as an argument:
```suggestion
    @classmethod
    def _generate_exception_args(cls):
        for objlen in (0, 1, 5):
            m = 2 * max(2, objlen)
            for start in range(-m, m):
                for end in range(-m, m):
                    yield objlen, start, end

    @classmethod
    def generate_encode_errors(cls):
        return tuple(
            UnsafeUnicodeEncodeError('utf-8', '0' * objlen, start, end, 'reason')
            for objlen, start, end in cls._generate_exception_args()
        )
```

Reply:
You're right. It could be easier like this.